php.buildEnv: Provide a `dev` output with utilities pointing to us

talyz opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
An attempt at a fix for #110826 and related issues. This makes `symlinkJoin` able to handle multiple outputs, a change which should probably be handled in a separate PR.

cc @mjsir911 @jcleng

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

talyz committed:
symlinkJoin: Add support for multiple outputs

Previously, `symlinkJoin` only cared about the main output of its
inputs. This makes it possible to specify which outputs we want to
link through the `outputs` parameter; if it's customized,
`symlinkJoin` will produce multiple outputs, each one the result of
combining the corresponding outputs of the packages listed in `paths`.

talyz committed:
php.buildEnv: Provide a `dev` output with utilities pointing to us

Create a customized `dev` output where the binaries provided point to
the wrapper, rather than the unwrapped php build.

mjsir911 commented:
This fixes #110826 for me, thanks a lot @talyz!

mjsir911 commented on pkgs/build-support/trivial-builders.nix:
```diff
@@ -318,23 +321,46 @@ rec {
     let
+      secondaryOutputs = builtins.filter (o: o != "out") outputs;
+      outputToPathsAttr = output: {
+        name = output + "Paths";
+        value = lib.flatten (map (p: p.${output} or []) paths);
+      };
       args = removeAttrs args_ [ "name" "postBuild" ]
         // {
           inherit preferLocalBuild allowSubstitutes;
-          passAsFile = [ "paths" ];
-        }; # pass the defaults
+          outputs = [ "out" ] ++ secondaryOutputs;
+          passAsFile = [ "paths" ] ++ map (o: o + "Paths") secondaryOutputs;
+        } // builtins.listToAttrs (map outputToPathsAttr secondaryOutputs);
     in runCommand name args
       ''
         mkdir -p $out
         for i in $(cat $pathsPath); do
           ${lndir}/bin/lndir -silent $i $out
         done
+
+        for output in ${toString secondaryOutputs}; do
+          eval output_path=\$$output
+          mkdir -p $output_path
+          for i in $(eval cat \$${output}PathsPath); do
+            ${lndir}/bin/lndir -silent $i $output_path
+
+            # Replace previously propagated build outputs with our
+            # main output
+            if [ -e $output_path/nix-support/propagated-build-inputs ]; then
+              rm $output_path/nix-support/propagated-build-inputs
+              echo "$out" >> $output_path/nix-support/propagated-build-inputs
+            fi
+          done
+        done
+
```
Is there a reason to treat `out` as special? Couldn't this logic be generalized to all outputs, whether `out` or `dev`?

talyz replied:
Well, there are two reasons:
1. The naming of the `paths` attribute doesn't match the naming scheme for the rest of the outputs (`*Paths`). This could be resolved by renaming it in the argument list to `runCommand`.
2. The additional outputs are treated specially in regards to propagated build outputs. If we want to generalize it, we would still need a check to make sure we don't mess with the main output's `propagated-build-inputs` file.

mjsir911 replied:
> The additional outputs are treated specially in regards to propagated build outputs. If we want to generalize it, we would still need a check to make sure we don't mess with the main output's propagated-build-inputs file.

I'm unfamiliar with the intricacies of this. Should `out` inherit the `paths`' propagated build inputs? Should the secondary outputs do the same, then? Do the secondary outputs have the original `out` path as a build input? Only the original `out` path, or could they have more?

> The naming of the paths attribute doesn't match the naming scheme for the rest of the outputs (*Paths). This could be resolved by renaming it in the argument list to runCommand.

Yeah, I think the generic approach would be simpler with this, no `secondaryOutputs`

```diff
diff --git a/pkgs/build-support/trivial-builders.nix b/pkgs/build-support/trivial-builders.nix
@@ -328,7 +328,6 @@ rec {
          , ...
          }:
     let
-      secondaryOutputs = builtins.filter (o: o != "out") outputs;
       outputToPathsAttr = output: {
         name = output + "Paths";
         value = lib.flatten (map (p: p.${output} or []) paths);
@@ -336,17 +335,11 @@ rec {
       args = removeAttrs args_ [ "name" "postBuild" ]
         // {
           inherit preferLocalBuild allowSubstitutes;
-          outputs = [ "out" ] ++ secondaryOutputs;
-          passAsFile = [ "paths" ] ++ map (o: o + "Paths") secondaryOutputs;
-        } // builtins.listToAttrs (map outputToPathsAttr secondaryOutputs);
+          passAsFile = map (o: o + "Paths") outputs;
+        } // builtins.listToAttrs (map outputToPathsAttr outputs);
     in runCommand name args
       ''
-        mkdir -p $out
-        for i in $(cat $pathsPath); do
-          ${lndir}/bin/lndir -silent $i $out
-        done
-
-        for output in ${toString secondaryOutputs}; do
+        for output in ${toString outputs}; do
           eval output_path=\$$output
           mkdir -p $output_path
           for i in $(eval cat \$${output}PathsPath); do
```


Thank you for your clarifications!

talyz commented on pkgs/build-support/trivial-builders.nix:
```diff
@@ -318,23 +321,46 @@ rec {
           inherit preferLocalBuild allowSubstitutes;
-          passAsFile = [ "paths" ];
-        }; # pass the defaults
+          outputs = [ "out" ] ++ secondaryOutputs;
```
This makes sure that `"out"` is always present in the list - the user could supply a list without it, but we always need it. I suppose we could just always add it and run `lib.unique` on the list instead.

mjsir911 replied:
I see! does it ever make sense to have an `outputs` provided to `symlinkJoin` that does not contain `"out"`?

Maybe it would make more sense to have `secondaryOutputs` be an argument to symlinkJoin, and you have a guarantee that you can't ever not include `out` in the outputs

talyz replied:
No, it wouldn't make sense, since we're using `out` as our main output. Since `output` is an argument to `mkDerivation` and the expected name of the parameter, I'd prefer to keep it and only use `secondaryOutputs` internally.

mjsir911 replied:
:+1: Sounds good. I can foresee some confusion by allowing an `output` without `"out"` (does `mkDerivation` allow the same?), what would the user even expect to happen? 

But it's such an edge case & doesn't make sense anyways, I think is good enough, if not a bit unintuitive.

Thank you for explaining!

talyz replied:
Yes, you don't need an `"out"` output, `mkDerivation` uses the first one in the list as the primary one. Some packages call their main output `"bin"` instead, for example. However, we need it, since we're using it to create a combined output of all main outputs (which can vary in name). So, unless we want to change that behavior, we should make sure we always get an `"out"` output.

jtojnar commented on pkgs/build-support/trivial-builders.nix:
```diff
@@ -318,23 +321,46 @@ rec {
         done
+
+        for output in ${toString secondaryOutputs}; do
+          eval output_path=\$$output
```
Rather then eval, you can use indirect variable access: `output_path=${!output}`: https://tldp.org/LDP/abs/html/bashver2.html#EX78

talyz replied:
done!

jtojnar commented on pkgs/build-support/trivial-builders.nix:
```diff
@@ -318,23 +321,46 @@ rec {
+            # Replace previously propagated build outputs with our
+            # main output
+            if [ -e $output_path/nix-support/propagated-build-inputs ]; then
+              rm $output_path/nix-support/propagated-build-inputs
```
This sounds quite destructive, the output might have been propagating something else than just `oldOut`. On the other hand, we probably want environments created by `symlinkJoin` to be self-contained so maybe it is okay that propagation is cut-off.

talyz replied:
Yes, it could be destructive, but for secondary outputs, I could only find `propagated-build-inputs` being added by the multiple outputs hook, not to say that it doesn't exist, but it doesn't seem very prevalent. Propagation shouldn't be cut off for the main output, at least. :)

jtojnar replied:
`propagatedBuildInputs` attribute [generates](https://github.com/NixOS/nixpkgs/blob/066fb63b0285cae5dc724ef2fd0b1f6c177e02cf/pkgs/stdenv/generic/setup.sh#L1155) the file in designated `dev` output (in the [order of preference](https://github.com/NixOS/nixpkgs/blob/066fb63b0285cae5dc724ef2fd0b1f6c177e02cf/pkgs/build-support/setup-hooks/multiple-outputs.sh#L32) `$outputDev`, `dev`, `out`).

talyz replied:
Ah, I see. I had missed that. Then it might be interesting to try copying and mangling the file to replace specifically the main output, but leave the others as-is.

talyz replied:
It now concatenates all the `paths`' `propagated-build-input` files per output, replacing the previous main output with `$out`. This should be much less destructive.

jtojnar commented on pkgs/build-support/trivial-builders.nix:
```diff
@@ -300,6 +300,9 @@ rec {
    *     |       `-- stack.fish -> /nix/store/6lzdpxshx78281vy056lbk553ijsdr44-stack-2.1.3.1/share/fish/vendor_completions.d/stack.fish
    * ...
    *
+   * # creates multiple outputs, each the result of combining the corresponding outputs of the paths
```
This comment is a bit imprecise as, like before, the `out` output will contain the primary output (first item in `outputs` list) when passed package attribute. Or even specific output when passed that.

Try:

```ShellSession
$ nix-build -E 'with import ./. {}; symlinkJoin { name = "foo"; paths = [ glib.dev ]; }'
$ nix-build -E 'with import ./. {}; symlinkJoin { name = "foo"; paths = [ glib ]; }'
```

talyz replied:
The comment refers to the example given just below it, not generally.

talyz replied:
Maybe that could be made more obvious, though. I was trying to follow the structure of the examples above it..

jtojnar replied:
Even then, you will end up with `openssl.bin` in `out`, not `openssl.out`.

talyz replied:
Ah, yes, now I see what you mean. I've updated the comment to reflect this.

I would guess that we want the main outputs to be combined like this almost always, but if there are times where it's inappropriate, we a parameter to change the behavior. Do you see a use-case where we need the outputs to correspond exactly?

talyz committed:
fixup! symlinkJoin: Add support for multiple outputs

talyz committed:
fixup! symlinkJoin: Add support for multiple outputs

jtojnar commented on pkgs/build-support/trivial-builders.nix:
```diff
@@ -300,6 +300,13 @@ rec {
+   * # of combining the main outputs of all packages in `paths`, and
+   * # `bin` and `dev` are the result of combining their corresponding
+   * # outputs of all packages in `paths`
+   * symlinkJoin { name = "multiexample"; paths = [ pkgs.libpng pkgs.openssl ]; outputs = [ "out" "bin" "dev" ]; }
```
Thinking about this more, this will probably only be useful if the secondary outputs are disjoint with the set of main outputs of the paths. Otherwise, you will end up with the same content in multiple outputs like here with `openssl.bin`.

talyz replied:
Yes, indeed. I don't think there's a "right answer" here, but I think the current one should cover most use-cases. In this case, you can ignore the `bin` output, since it's more likely you want the main output of all paths, but if you really have a need for it, it's available. Or do you have a better idea of how this should work?

jtojnar replied:
Yeah, I do not think there is right answer either. But we probably should not include things like this in an example – I would either remove the `bin` output or use `openssl.out`. And add a comment emphasizing that behaviour.

talyz replied:
I've updated the example a bit, removing the `bin` output and adding `man` instead. I also added a comment about explicitly using `<pkg>.out` for packages where it's not the main output. I guess an alternative to this would be to call our main output something different, like `all_mains`, and let the others be pure summations of their respective outputs.

jtojnar replied:
Hmm, actually that raises another question. If one explicitly adds `openssl.out` should `openssl.dev` (actually `openssl.out.dev`) still end up in dev? Or should we use `lib.optionalString ((path.outputUnspecified or false) && path ? ${output})`.

talyz replied:
Hm, yeah, that's a good question. I'm leaning towards the latter, but I'm not really sure..

talyz committed:
fixup! php.buildEnv: Provide a `dev` output with utilities pointing to us

jtojnar commented on pkgs/build-support/trivial-builders.nix:
```diff
@@ -300,6 +300,18 @@ rec {
+   *
+   * Note that this means that if a package uses `out` as a secondary
+   * output, you need to explicitly include it if you're interested in
+   * its contents. For example, if you want the `openssl`s `out`
```
```suggestion
   * its contents. For example, if you want the `openssl`’s `out`
```

mjsir911 commented:
I'm getting this error with the latest:
```nix
(pkgs.php80.withExtensions ({all, ...}: with all; [tokenizer]))
/nix/store/2j6ikpjsd8iz3kygdm3nqw821k16b8i3-php-with-extensions-8.0.5.drv
```

```sh
$ nix build /nix/store/2j6ikpjsd8iz3kygdm3nqw821k16b8i3-php-with-extensions-8.0.5.drv
builder for '/nix/store/2j6ikpjsd8iz3kygdm3nqw821k16b8i3-php-with-extensions-8.0.5.drv' failed with exit code 1; last 1 log lines:
  /nix/store/xz78aaqmmhdfrmghj0zwb7fykax3a09x-hook/nix-support/setup-hook: line 109: flagsBefore: unbound variable
[1 built (1 failed), 88 copied (204.6 MiB), 42.4 MiB DL]
error: build of '/nix/store/2j6ikpjsd8iz3kygdm3nqw821k16b8i3-php-with-extensions-8.0.5.drv' failed
```

talyz committed:
fixup! symlinkJoin: Add support for multiple outputs

talyz commented:
@mjsir911 Oops, try it now.

mjsir911 commented:
Fixes it, ty @talyz

talyz committed:
php.extensions: Use a PHP package with the required internalDeps

When building an extension, provide a PHP with the prerequisite
extensions loaded. This means more tests succeed / aren't skipped
because the extension fails to load.

talyz committed:
php.extensions: Don't hog pre- and postPatch in mkExtension

talyz committed:
fixup! symlinkJoin: Add support for multiple outputs

mjsir911 commented on pkgs/top-level/php-packages.nix:
```diff
@@ -212,12 +215,25 @@ lib.makeScope pkgs.newScope (self: with self; {
 
       enableParallelBuilding = true;
-      nativeBuildInputs = [ php.unwrapped autoconf pkg-config re2c ];
+      nativeBuildInputs = [
```
This change would be useful in `build-pecl.nix` for me as well! Not sure if it would apply generally to all `buildPecl` built packages, but right now I have to specify: 
```
nativeBuildInputs = [ pkg-config libgit2 (php.withExtensions ({all, ...}: with all; [tokenizer])) ];
```

for my own package.

Setting `internalDeps` doesn't seem to do anything with `buildPecl`. How am I supposed to use the `ext` directory?

mjsir911 commented:
@talyz Anything to be done to move this along? I've been using it successfully with 
```
let php=(import (builtins.fetchTarball https://github.com/talyz/nixpkgs/archive/refs/heads/php-dev-output.tar.gz) {}).php80;
```
for a while now but would love to see this in mainline soon

jtojnar commented:
To move this forward, it might be a good idea to have a PR just for the `symlinkJoin` change and invite wider community for input. There are still some rough edges and  `symlinkJoin` is quite a foundational function so the review should match that.

mjsir911 commented:
@talyz not sure about getting this merged in but could you rebase onto latest? I've been depending on your fork for part of my setup.
